### ISSUE TITLE
Provider response: NoSuchInstance vs NoSuchObject when given an invalid instance ID?

### ISSUE THREAD

**Issue by thom-nic** [markabrahams/node-net-snmp#259]
Hi Mark!  

Our table handlers add and remove rows using `mib.addTableRow(TABLE_NAME, [instanceID, ...values])`. I've just noticed that when a client requests a table OID for an instance that does not exist, we get a `noSuchObject` response.  I would have expected `noSuchInstance` instead.  

I can get a response for the same OID with a valid instance just to confirm the table is set up correctly.  Also I've tested this as a v1 and v2c request with the same results.

Is noSuchObject the correct response for a tabular OID whose instance does not exist? Or am I doing something wrong?

**Comment by markabrahams** [comment]
Hi Thom

The RFC portion that specifies the behaviour here is RFC 3416 Section 4.2.1:

4.2.1.  The GetRequest-PDU

   A GetRequest-PDU is generated and transmitted at the request of an
   application.

   Upon receipt of a GetRequest-PDU, the receiving SNMP entity processes
   each variable binding in the variable-binding list to produce a
   Response-PDU.  All fields of the Response-PDU have the same values as
   the corresponding fields of the received request except as indicated
   below.  Each variable binding is processed as follows:

   (1)   If the variable binding's name exactly matches the name of a
         variable accessible by this request, then the variable
         binding's value field is set to the value of the named
         variable.

   (2)   Otherwise, if the variable binding's name does not have an
         OBJECT IDENTIFIER prefix which exactly matches the OBJECT
         IDENTIFIER prefix of any (potential) variable accessible by
         this request, then its value field is set to "noSuchObject".

   (3)   Otherwise, the variable binding's value field is set to
         "noSuchInstance".

Interpreting condition (2) above *seems* to be supporting the use of
NoSuchInstance in this case, although I've read it five times now, and
don't think it's sufficiently specific (the use of "potential" and "this
request" phrases are both open to interpretation).

If this case does fail condition (2), and there is definitely a good case
for this, then NoSuchInstance should be the reply.

Since this also matches my general Intuition, it seems that the library
should indeed in this case return NoSuchInstance instead of NoSuchObject.

I'll have a look at making this change - although that piece of the library
code might need some rework, given the current nature of the tree-based
processing.

Regards, Mark.

**Comment by thom-nic** [comment]
Hey thanks Mark.  Single anecdotal data point but I did a quick test on an APC UPS which we have here:

```
$ snmpwalk -v2c -c private -m ./powernet441.mib eng-apc-ups-01 upsOutletGroupStatusGroupState 
PowerNet-MIB::upsOutletGroupStatusGroupState.1 = INTEGER: upsOutletGroupStatusOn(1)
PowerNet-MIB::upsOutletGroupStatusGroupState.2 = INTEGER: upsOutletGroupStatusOn(1)
PowerNet-MIB::upsOutletGroupStatusGroupState.3 = INTEGER: upsOutletGroupStatusOn(1)

$ snmpget -v2c -c private -m ./powernet441.mib eng-apc-ups-01 upsOutletGroupStatusGroupState.3
PowerNet-MIB::upsOutletGroupStatusGroupState.3 = INTEGER: upsOutletGroupStatusOn(1)

$ snmpget -v2c -c private -m ./powernet441.mib eng-apc-ups-01 upsOutletGroupStatusGroupState.4
PowerNet-MIB::upsOutletGroupStatusGroupState.4 = No Such Instance currently exists at this OID

$ snmpget -v2c -c private -m ./powernet441.mib eng-apc-ups-01 rPDUOutletStatus.1              
PowerNet-MIB::rPDUOutletStatusTable = No Such Object available on this agent at this OID
```

When I give an instance that doesn't appear in the walk (third command) I get "no such instance"

When I give an OID for a table that doesn't exist at all for that product, I get "No such object" (fourth command)

**Comment by thom-nic** [comment]
Hi Mark - just checking to see if this looks like something that can be addressed.  I looked at the code and think I found the relevant lines here: https://github.com/markabrahams/node-net-snmp/blob/master/index.js#L4991-L5016

I'm sure I don't completely understand what's going on, but it seems like the pseudocode should be something like

```
If I don't have any column handler for this OID:
    return noSuchObject

else if I don't have a row in the table for this instance:
    return noSuchInstance
```

**Comment by markabrahams** [comment]
Hi Thom - I've push an update for this now, available in version 3.15.2 of the npm.

**Comment by thom-nic** [comment]
Confirmed fixed in 07c119f - thanks @markabrahams!
